core/new-gui/src/app/workspace/service/joint-ui/joint-ui.service.ts

```diff
@@ -0,0 +1,253 @@
+   * This function takes an source and target OperatorPort and
+   * creates a JointJS link element that would be added to the JointJS
+   * graph to let JointJS display the operator accordingly. This function
+   * will connect the source and the target operator together on the graph.
+   *
+   * @param source the OperatorPort of the source of a link
+   * @param target the OperatorPort of the target of a link
+   * @returns JointJS Link Element
+   */
+  public getJointjsLinkElement(
+    source: OperatorPort, target: OperatorPort
+  ): joint.dia.Link {
+    const link = this.getDefaultLinkElement();
+    link.set('source', { id: source.operatorID, port: source.portID });
+    link.set('target', { id: target.operatorID, port: target.portID });
+    return link;
+  }
+
+  /**
+   * This function will creates a custom JointJS link element using
+   * custom attributes / styles to display the operator. This function
+   * defines the svg properties for each part of link, such as the
+   * shape of the arrow or the link. Other styles are defined in the
+   * "app/workspace/component/workflow-editor/workflow-editor.component.scss".
+   * The reason for separating styles in svg and css is that while we can
+   * change the shape of the operators in svg, according to JointJS official
+   * website, https://resources.jointjs.com/tutorial/element-styling, "
+   * CSS properties have higher precedence over SVG element attributes."
+   * As a result, a separate css/scss file is required to override the default
+   * style of the operatorLink.
+   * @returns JointJS Link
+   */
+  public getDefaultLinkElement(): joint.dia.Link {
+    const link = new joint.dia.Link({
+      attrs: {
+        '.connection-wrap': {
+          'stroke-width': 0
+        },
+        '.marker-source': {
+          d: this.sourceOperatorHandle,
+          stroke: 'none',
+          fill: '#919191'
+        },
+        '.marker-arrowhead-group-source .marker-arrowhead': {
+          d: this.sourceOperatorHandle,
+        },
+        '.marker-target': {
+          d: this.targetOperatorHandle,
+          stroke: 'none',
+          fill: '#919191'
+        },
+        '.marker-arrowhead-group-target .marker-arrowhead': {
+          d: this.targetOperatorHandle,
+        },
+        '.tool-remove': {
+          fill: '#D8656A',
+          width: 24
+        },
+        '.tool-remove path': {
+          d: this.deleteButtonPath,
+        },
+        '.tool-remove circle': {
+
+          // visibility: 'hidden'
+        }
+      }
+    });
+    return link;
+  }
+
+  /**
+   * This function registers a custom JointJS shape in the
+   * joint.shapes.devs map so it can be used later on.
+   */
+  private setupCustomJointjsModel(): void {
+    joint.shapes.devs['TexeraOperatorShape'] = joint.shapes.devs.Model.extend({
+      type: 'devs.TexeraModel',
+      markup:
+        `<g class="element-node">
+          <rect class="body" stroke-width="2" stroke="blue" rx="5px" ry="5px"></rect>
```

Comment:
Is this just to add a custom delete button to the operator's jointjs model and other properties of the model are being set in getJointjsOperatorElement()?

Reply:
Yes, setupCustomJointjsModel() basically creates an empty rectangle with delete button on the top-right hand corner. Other properties such as styles and texts are set in the getJointJsOperatorElement(). I will modify the comments ASAP.